tensorflow-training/src/main/java/org/tensorflow/training/examples/MNISTTest.java

```diff
@@ -0,0 +1,331 @@
+
+    // Flatten inputs
+    Reshape<TFloat32> flatten = tf.reshape(pool2, tf.concat(Arrays.asList(tf.slice(tf.shape(pool2), tf.constant(new int[]{0}), tf.constant(new int[]{1})), tf.constant(new int[]{-1})), tf.constant(0)));
+
+    // Fully connected layer
+    Variable<TFloat32> fc1Weights = tf.variableWithInit(tf.math.mul(tf.random.truncatedNormal(tf.constant(new int[]{IMAGE_SIZE*IMAGE_SIZE*4,512}), TFloat32.DTYPE, TruncatedNormal.seed(SEED)), tf.constant(0.1f)));
+    Variable<TFloat32> fc1Biases = tf.variableWithInit(tf.fill(tf.constant(new int[]{512}),tf.constant(0.1f)));
+    Relu<TFloat32> relu3 = tf.nn.relu(tf.math.add(tf.linalg.matMul(flatten, fc1Weights), fc1Biases));
+
+    // Softmax layer
+    Variable<TFloat32> fc2Weights = tf.variableWithInit(tf.math.mul(tf.random.truncatedNormal(tf.constant(new int[]{512,NUM_LABELS}), TFloat32.DTYPE, TruncatedNormal.seed(SEED)), tf.constant(0.1f)));
+    Variable<TFloat32> fc2Biases = tf.variableWithInit(tf.fill(tf.constant(new int[]{NUM_LABELS}),tf.constant(0.1f)));
+
+    Add<TFloat32> logits = tf.math.add(tf.linalg.matMul(relu3, fc2Weights), fc2Biases);
+
+    // Predicted outputs
+    Softmax<TFloat32> prediction = tf.withName(OUTPUT_NAME).nn.softmax(logits);
+
+    // Loss function & regularization
+    OneHot<TFloat32> oneHot = tf.oneHot(labels, tf.constant(10), tf.constant(1.0f), tf.constant(0.0f));
+    SoftmaxCrossEntropyWithLogits<TFloat32> batchLoss = tf.nn.softmaxCrossEntropyWithLogits(logits, oneHot);
+    Mean<TFloat32> labelLoss = tf.math.mean(batchLoss.loss(), tf.constant(0));
+    Add<TFloat32> regularizers = tf.math.add(tf.nn.l2Loss(fc1Weights), tf.math.add(tf.nn.l2Loss(fc1Biases), tf.math.add(tf.nn.l2Loss(fc2Weights), tf.nn.l2Loss(fc2Biases))));
+    Add<TFloat32> loss = tf.withName(TRAINING_LOSS).math.add(labelLoss, tf.math.mul(regularizers, tf.constant(5e-4f)));
+
+    optimizerName = optimizerName.toLowerCase();
+    // Optimizer
+    Optimizer optimizer;
+    switch (optimizerName) {
+      case "adadelta":
+        optimizer = new AdaDelta(graph, 1f, 0.95f, 1e-8f);
+        break;
+      case "adagradda":
+        optimizer = new AdaGradDA(graph, 0.01f);
+        break;
+      case "adagrad":
+        optimizer = new AdaGrad(graph, 0.01f);
+        break;
+      case "adam":
+        optimizer = new Adam(graph,0.001f,0.9f,0.999f,1e-8f);
+        break;
+      case "sgd":
+        optimizer = new GradientDescent(graph,0.01f);
+        break;
+      case "momentum":
+        optimizer = new Momentum(graph, 0.01f, 0.9f, false);
+        break;
+      case "rmsprop":
+        optimizer = new RMSProp(graph,0.01f, 0.9f, 0.0f, 1e-10f, false);
+        break;
+      default:
+        throw new IllegalArgumentException("Unknown optimizer " + optimizerName);
+    }
+    logger.info("Optimizer = " + optimizer.toString());
+    Op minimize = optimizer.minimize(loss, TRAIN);
+
+    Op init = graph.variablesInitializer();
```

Comment:
Without calling `graph.variablesInitializer()` it doesn't actually create the init node on the graph, as it doesn't know when you've finished adding nodes. Similar to the minimize call above, these both mutate the graph to add appropriate entry points, but I don't need the ops they produce here (because they get special names).